### PR TITLE
Implement a hook in base for disabling threading

### PR DESCRIPTION
**Pull request by fredrikekre**
Implement a hook in base for disabling threading of libraries, use it to decouple `Distributed` from `LinearAlgebra`.

Alternative to #29975, WIP because I don't know if this is the interface we want, where the code should live, and what the names of the functions should be. Maybe not very important since these are all internal functions.

### CODE REVIEW

**Comment by fredrikekre** [comment]
Bump for Julia 1.1

File: base/initdefs.jl [hunk 1]
```diff
@@ -237,3 +237,29 @@ function _atexit()
...
+end
+
+function disable_library_threading()
+    global library_threading_enabled = false
+    while !isempty(disable_library_threading_hooks)
+        f = pop!(disable_library_threading_hooks)
+        try
+            f()
+        catch err
+            print(stderr, "ERROR: "); showerror(stderr, err); println(stderr)
```
Comment (vtjnash) [inline base/initdefs.jl]: use `@warn "some descriptive text" exception=ex,catch_backtrace()`

**Comment by StefanKarpinski** [comment]
Triage is in favor if you want to get this passing CI and merge it. Wants a NEWS entry too, right?

**Review by JeffBezanson — APPROVED** [review]

**Comment by fredrikekre** [comment]
Great, updated. The last CI runs was during some flaky times IIRC, so I don't think there was something wrong with the PR.

**Comment by ararslan** [comment]
This seems to be causing the Distributed tests to fail on PkgEval:
```
ERROR: LoadError: On worker 2:
ArgumentError: Package LinearAlgebra not found in current path:
- Run `import Pkg; Pkg.add("LinearAlgebra")` to install the LinearAlgebra package.
```

**Pull request merged**